addon_service/credentials/encryption.py

```diff
@@ -0,0 +1,142 @@
+class KeyParameters:  # https://datatracker.ietf.org/doc/html/rfc7914#section-2
+    salt: bytes = dataclasses.field(default_factory=salt_factory)
+    # recommended scrypt_cost ("N") between 2^14 and 2^20
+    scrypt_cost: int = settings.GRAVYVALET_SCRYPT_COST or 2**17
+    # recommended scrypt_block_size ("r") = 8
+    scrypt_block_size: int = settings.GRAVYVALET_SCRYPT_BLOCK_SIZE or 8
+    # recommended scrypt_parallelization ("p") = 1
+    scrypt_parallelization: int = settings.GRAVYVALET_SCRYPT_PARALLELIZATION or 3
+
+    def __post_init__(self):
+        assert self.scrypt_block_size > 1
+        # scrypt_cost "must be larger than 1, a power of 2, and less than 2^(128 * r / 8)"
+        _cost_log2 = math.log2(self.scrypt_cost)
+        assert (
+            self.scrypt_cost > 1
+            and _cost_log2 == int(_cost_log2)
+            and _cost_log2 <= (128 * self.scrypt_block_size / 8)
+        )
+        # scrypt_parallelization "is a positive integer less than or equal to ((2^32-1) * 32) / (128 * r)"
+        assert self.scrypt_parallelization <= ((2**32 - 1) * 32) / (
+            128 * self.scrypt_block_size
+        )
+
+    def memory_required(self) -> int:
+        return self.scrypt_cost * self.scrypt_block_size * 129
+
+
+def pls_encrypt_json(jsonable_obj, key_params: KeyParameters) -> bytes:
+    return pls_encrypt_bytes(json.dumps(jsonable_obj).encode(), key_params)
+
+
+def pls_decrypt_json(encrypted_json: bytes, key_params: KeyParameters):
+    return json.loads(pls_decrypt_bytes(encrypted_json, key_params))
+
+
+def pls_encrypt_bytes(msg: bytes, key_params: KeyParameters) -> bytes:
+    return _derive_multifernet_key(key_params).encrypt(msg)
+
+
+def pls_decrypt_bytes(encrypted: bytes, key_params: KeyParameters) -> bytes:
+    return _derive_multifernet_key(key_params).decrypt(encrypted)
+
+
+def pls_rotate_encryption(
+    encrypted: bytes,
+    stored_params: KeyParameters,
+) -> tuple[bytes, KeyParameters]:
+    _fresh_params = KeyParameters(salt=salt_factory())
+    _stored_up_to_date = (
+        len(stored_params.salt) == len(_fresh_params.salt)
+        and stored_params.scrypt_cost == _fresh_params.scrypt_cost
+        and stored_params.scrypt_block_size == _fresh_params.scrypt_block_size
+        and stored_params.scrypt_parallelization == _fresh_params.scrypt_parallelization
+    )
+    if _stored_up_to_date:  # key params NOT changed -- can use MultiFernet.rotate
+        _fresh_encrypted = _derive_multifernet_key(stored_params).rotate(encrypted)
+        _fresh_params = stored_params
+    else:
+        # key defaults HAVE changed -- decrypt and re-encrypt
+        _decrypted = pls_decrypt_bytes(encrypted, stored_params)
+        _fresh_encrypted = pls_encrypt_bytes(_decrypted, _fresh_params)
+    return _fresh_encrypted, _fresh_params
+
+
+# deriving keys is expensive on purpose -- cache, but only in local memory
+@functools.lru_cache(maxsize=settings.GRAVYVALET_DERIVED_KEY_CACHE_SIZE)
```

Comment:
This is a nitpick. But should `GRAVYVALET_DERIVED_KEY_CACHE_SIZE` be a calculated value instead of a constant? I assume the size of the key generated depends on the parameters used. So it might not always make sense to have a constant value for the cache, but a variable value based on `memory_required`?

Reply:
the size of the derived key is a constant 32 bytes (as required by [cryptography.fernet.Fernet](https://cryptography.io/en/latest/fernet/#cryptography.fernet.Fernet) and set in `_KEY_BYTE_COUNT`) -- scrypt's parameters affect the amount of memory required for scrypt to run, but that memory can be (is?) freed as soon as it's done